yt_dlp/extractor/youtube.py

```diff
@@ -2643,30 +2643,51 @@ def _extract_signature_timestamp(self, video_id, player_url, ytcfg=None, fatal=F
         return sts
 
     def _mark_watched(self, video_id, player_responses):
-        playback_url = get_first(
-            player_responses, ('playbackTracking', 'videostatsPlaybackUrl', 'baseUrl'),
-            expected_type=url_or_none)
-        if not playback_url:
+        mark_watched_urls = {
+            "playback": get_first(
+                player_responses, ('playbackTracking', 'videostatsPlaybackUrl', 'baseUrl'),
+                expected_type=url_or_none),
+            "watchtime": get_first(
+                player_responses, ('playbackTracking', 'videostatsWatchtimeUrl', 'baseUrl'),
+                expected_type=url_or_none)
+        }
+
+        if not all(mark_watched_urls.values()):
             self.report_warning('Unable to mark watched')
             return
-        parsed_playback_url = compat_urlparse.urlparse(playback_url)
-        qs = compat_urlparse.parse_qs(parsed_playback_url.query)
-
-        # cpn generation algorithm is reverse engineered from base.js.
-        # In fact it works even with dummy cpn.
-        CPN_ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_'
-        cpn = ''.join(CPN_ALPHABET[random.randint(0, 256) & 63] for _ in range(0, 16))
-
-        qs.update({
-            'ver': ['2'],
-            'cpn': [cpn],
-        })
-        playback_url = compat_urlparse.urlunparse(
-            parsed_playback_url._replace(query=compat_urllib_parse_urlencode(qs, True)))
-
-        self._download_webpage(
-            playback_url, video_id, 'Marking watched',
-            'Unable to mark watched', fatal=False)
+        for url_type, url in mark_watched_urls.items():
```

Comment:
```py
        for is_full, key in enumerate(('videostatsPlaybackURL' 'videostatsWatchtimeURL')):
            label = 'fully ' if is_full else ''
            url = get_first(player_responses, ('playbackTracking', key, 'baseUrl'),
                            expected_type=url_or_none)
            if not url:
                self.report_warning(f'Unable to mark {label}watched')
                return
```
This way, if watchtime URL extraction fails, the partial marking still runs

PS: label can be used in the `_download_webpage` message too

Reply:
👍